File: bench/cat.go
// BOT-GENERATED

package main

import (
	"fmt"
	"io"
	"os"
)

func copyInput(input *os.File, name string) error {
	if _, err := io.Copy(os.Stdout, input); err != nil {
		return fmt.Errorf("%s: %w", name, err)
	}
	return nil
}

func copyFile(name string) error {
	if name == "-" {
		return copyInput(os.Stdin, "stdin")
	}

	input, err := os.Open(name)
	if err != nil {
		return err
	}

	copyErr := copyInput(input, name)
	closeErr := input.Close()
	if copyErr != nil {
		return copyErr
	}
	return closeErr
}

func run(args []string) error {
	if len(args) == 0 {
		return copyInput(os.Stdin, "stdin")
	}

	for _, name := range args {
		if err := copyFile(name); err != nil {
			return err
		}
	}
	return nil
}

func main() {
	if err := run(os.Args[1:]); err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
}

File: bench/cat.c
// BOT-GENERATED

#include <errno.h>
#include <stdio.h>
#include <string.h>

static int copy(FILE *inp, const char *name) {
  unsigned char buf[64 * 1024];
  size_t        len;

  while ((len = fread(buf, 1, sizeof(buf), inp))) {
    if (fwrite(buf, 1, len, stdout) != len) {
      perror("stdout");
      return 1;
    }
  }

  if (ferror(inp)) {
    perror(name);
    return 1;
  }
  return 0;
}

int main(int argc, char **argv) {
  if (argc == 1) {
    if (copy(stdin, "stdin")) return 1;
  }

  for (int ind = 1; ind < argc; ind++) {
    if (!strcmp(argv[ind], "-")) {
      if (copy(stdin, "stdin")) return 1;
      continue;
    }

    FILE *input = fopen(argv[ind], "rb");
    if (!input) {
      perror(argv[ind]);
      return 1;
    }
    const int failed = copy(input, argv[ind]);
    if (fclose(input) && !failed) {
      perror(argv[ind]);
      return 1;
    }
    if (failed) return 1;
  }

  if (fflush(stdout)) {
    perror("stdout");
    return 1;
  }
  return 0;
}
